packages/wallet/src/SingleAddressWallet/SingleAddressWallet.ts

```diff
@@ -255,13 +266,17 @@ export class SingleAddressWallet implements ObservableWallet {
         keyAgent.knownAddresses$.pipe(
           distinctUntilChanged(groupedAddressesEquals),
           tap(
-            // derive an address if none available
+            // derive addresses if none available
             (addresses) => {
               if (addresses.length === 0) {
-                this.#logger.debug('No addresses available; deriving one');
-                void keyAgent
-                  .deriveAddress({ index: 0, type: AddressType.External }, 0)
-                  .catch(() => this.#logger.error('Failed to derive address'));
+                this.#logger.debug('No addresses available; initiating address discovery process');
+
+                void this.#addressDiscovery
+                  .discover(keyAgent, this.chainHistoryProvider, this.#addressLookAheadSearch)
+                  // The key agent 'remembers' every address it derives, however, some derived addresses
+                  // we don't want to remember (I.E gap addresses). So we must override the known addresses
+                  // with the curated list of addresses returned by the discovery process.
+                  .catch(() => this.#logger.error('Failed to complete the address discovery process'));
```

Comment:
If this fails, then the wallet is essentially bricked. We should consider a retry strategy (either here or as part of `discover` implementation).

Reply:
We typically use exponential backoff via [coldObservableProvider](https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/wallet/test/services/util/coldObservableProvider.test.ts) util. You could either use that (by converting promise-observable-promise), or find something else. I'd say cut it in like 5 minutes?

Reply:
Looks good, but would be nice to validate with test(s): create a wallet with mock AddressDiscovery that fails once or twice before resolving.